File: 5_kyu/PaginationHelper.java
import java.util.List;

public class PaginationHelper<I> {
    List<I> collection;
    int itemsPerPage;

  public PaginationHelper(List<I> collection, int itemsPerPage) {
    this.collection = collection;
    this.itemsPerPage = itemsPerPage;
  }
  
  public int itemCount() {
    return collection.size();
  }
  
  public int pageCount() {
    int division = (int)(collection.size() / itemsPerPage);
    int resto = collection.size() % itemsPerPage;
    return resto == 0 ? division : division + 1;
  }
  
  public int pageItemCount(int pageIndex) {
    int comienzoPagIndex = itemsPerPage * pageIndex;
    int finalPageIndex = comienzoPagIndex + itemsPerPage;

    if (comienzoPagIndex >= collection.size() || comienzoPagIndex < 0) {
        return -1;
    }else if (finalPageIndex >= collection.size()) {
        return collection.size() - comienzoPagIndex;
    } else {
        return itemsPerPage;
    }
  }
  
  public int pageIndex(int itemIndex) {
    return itemIndex >= collection.size() || itemIndex < 0 ? -1 : (int)(itemIndex / itemsPerPage);
  }
}

File: 5_kyu/PaginationHelper.js
class PaginationHelper {
    collection;
    itemsPerPage;

    constructor(collection, itemsPerPage) {
        this.collection = collection;
        this.itemsPerPage = itemsPerPage;
    }

    itemCount() {
        return this.collection.length;
    }

    pageCount() {
        let division = Math.floor(this.collection.length / this.itemsPerPage);
        let resto = this.collection.length % this.itemsPerPage;
        
        return resto === 0 ? division : division + 1;
    }

    pageItemCount(pageIndex) {
        let comienzoPagIndex =  this.itemsPerPage * pageIndex;
        let finalPagIndex = comienzoPagIndex + this.itemsPerPage;

        if (comienzoPagIndex >= this.collection.length || comienzoPagIndex < 0) {
            return -1;
        }else if (finalPagIndex >= this.collection.length) {
            return this.collection.length - comienzoPagIndex;
        } else {
            return this.itemsPerPage;
        }
    }

    pageIndex(itemIndex) {
        return itemIndex >= this.collection.length || itemIndex < 0 ? -1 : Math.floor(itemIndex / this.itemsPerPage);

    }
}
